c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1763,6 +2135,146 @@ class DeviceCachingAllocator {
     return blocks;
   }
 
+  // returns the smallest possible address in any segment
+  // where there is enough free address space to fit size
+  // may be composed of free and unmapped segments
+  Block* find_expandable_block(
+      int device,
+      cudaStream_t stream,
+      BlockPool* pool,
+      size_t size) {
+    Block key(device, stream, 0);
+
+    auto allocateable = [](Block* b) {
+      return b && !b->allocated && b->event_count == 0 &&
+          b->stream_uses.empty();
+    };
+    auto has_available_address_space = [&](Block* b) {
+      size_t bytes = 0;
+      while (bytes < size && allocateable(b)) {
+        bytes += b->size;
+        b = b->next;
+      }
+      return bytes >= size;
+    };
+    for (auto it = pool->unmapped.lower_bound(&key);
+         it != pool->unmapped.end() && (*it)->stream == stream;
+         ++it) {
+      Block* c = *it;
+      // we found the lowest address of an unmapped segment
+      // but there might be a free segment we can also use
+      // right before it
+      if (allocateable(c->prev)) {
+        c = c->prev;
+      }
+      if (has_available_address_space(c)) {
+        return c;
+      }
+    }
+    auto segment_size = pool->is_small ? kSmallBuffer : kLargeBuffer;
+    expandable_segments_.emplace_back(new ExpandableSegment(
+        device, stream, segment_size, devices_with_peer_access_));
+
+    ExpandableSegment* es = expandable_segments_.back();
+    Block* candidate = new Block(device, stream, es->size(), pool, es->ptr());
+    candidate->mapped = false;
+    candidate->expandable_segment_ = es;
+    pool->unmapped.insert(candidate);
+    return candidate;
+  }
+
+  bool map_block(
+      Block* to_map,
+      size_t size,
+      const std::shared_ptr<GatheredContext>& ctx) {
+    TORCH_INTERNAL_ASSERT(!to_map->mapped && size <= to_map->size);
+    TORCH_INTERNAL_ASSERT(
+        !to_map->history); // unmapped blocks should not keep history
+    auto mapped_range =
+        to_map->expandable_segment_->map(SegmentRange{to_map->ptr, size});
+    // failed to map the memory
+    if (mapped_range.size == 0) {
+      return false;
+    }
+    TORCH_INTERNAL_ASSERT(
+        mapped_range.ptr == to_map->ptr && mapped_range.size >= size);
+
+    BlockPool& pool = *to_map->pool;
+    pool.unmapped.erase(to_map);
+    to_map->mapped = true;
+
+    if (mapped_range.size < to_map->size) {
```

Comment:
Should we be calling `should_split` here ?

Reply:
That is a good question. I think with expandable segments we should _always_ split even if the free space is smaller than our normal split size because we might expand the segment later and wouldn't want to lose this memory. This means that should_split maybe should always be true for the normal allocator parts as well (though we are talking about losing 512 bytes the way it is currently set up).